File: 백준/Gold/12967. pqr/pqr.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        long result = 0;

        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        int[] nums = new int[N];
        int[] sub = new int[1000001];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            nums[i] = Integer.parseInt(st.nextToken());

            int divide = 1;
            int cur = nums[i];

            while (divide <= (int) Math.sqrt(cur)) {
                if (cur % divide == 0) {
                    sub[divide]++;

                    int divided = cur / divide;

                    if (divided != divide && divided <= 1000000) {
                        sub[cur/divide]++;
                    }
                }
                divide++;
            }
        }

        for (int i = 0; i < N; i++) {
            long n1 = nums[i];
            for (int j = i+1; j < N; j++) {
                long n2 = nums[j];

                int y = (int) (K / gcd(n1 *n2, K));

                result += sub[y];

                if (n1 % y == 0) {
                    result--;
                }

                if (n2 % y == 0) {
                    result--;
                }
            }
        }

        System.out.println(result/3);
    }

    static long gcd(long a, long b) {

        if (b == 0) {
            return a;
        }

        return gcd(b, a % b);
    }
}

File: 백준/Gold/12967. pqr/pqr.py
# pqr
import sys
input = sys.stdin.readline

def gcd(a, b):
    if b > a:
        a, b = b, a

    while b:
        a, b = b, a % b
        
    return a

N, K = map(int,input().split())
nums = list(map(int,input().split()))
result = 0

sub = [0] * 1000001

for i in range(N):
    divide = 1
    cur = nums[i]
    
    while divide <= int(cur**0.5):
        if cur % divide == 0:
            sub[divide] += 1
            divided = cur // divide

            if divided != divide and divided <= 1000000:
                sub[cur//divide] += 1

        divide += 1

for i in range(N):
    n1 = nums[i]
    for j in range(i+1, N):
        n2 = nums[j]
        y = K // gcd(n1*n2, K)

        result += sub[y]
        
        if n1 % y == 0:
            result -= 1

        if n2 % y == 0:
            result -= 1
            
print(result // 3)
